### PR TITLE
apps/ble_perfs: Implement BLE Server to measure the ble bandwidth

### PR DESCRIPTION
**Pull request by amitpurwar**
Signed-off-by: Amit Purwar <amit.purwar@samsung.com>

### CODE REVIEW

File: apps/examples/ble_perfs/Make.defs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+###########################################################################
+#
+# Copyright 2021 Samsung Electronics All Rights Reserved.
```
Comment (jeongchanKim) [inline apps/examples/ble_perfs/Make.defs:3]: Please change the license year to 2022.

File: apps/examples/ble_perfs/Makefile [hunk 2]
```diff
@@ -0,0 +1,122 @@
+###########################################################################
+#
+# Copyright 2021 Samsung Electronics All Rights Reserved.
```
Comment (jeongchanKim) [inline apps/examples/ble_perfs/Makefile:3]: Same as above.

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+
+static int perfs_ble_deinit(void)
+{
+	ble_result_e ret = BLE_MANAGER_FAIL;
+	wifi_manager_result_e res = WIFI_MANAGER_SUCCESS;
+
+	res = wifi_manager_deinit();
+	if (res != WIFI_MANAGER_SUCCESS) {
+		BLE_LOGD( "Fail to deinit wifi manager [%d] \n", res);
+		return -1;
```
Comment (jeongchanKim) [inline apps/examples/ble_perfs/ble_perfs_main.c:374]: Is it okay not to deinit the ble manager in here?

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+
+        BLE_LOGD( "diff time: %ld sec %ld usec \n", diff_time.tv_sec, diff_time.tv_usec);
+	
+	total_time = diff_time.tv_sec + (double) diff_time.tv_usec / 1000000;
+	speed = g_perf_data.packet_size * g_perf_data.packet_count * 8 / total_time;
+
+	BLE_LOGD( "=================================================\n");
+	BLE_LOGD( "Packet Count  : % d \n", g_perf_data.packet_count);
+	BLE_LOGD( "Packet Size   : % d byte\n", g_perf_data.packet_size);
+	BLE_LOGD( "Total time    : %.2lf ms \n", total_time*1000);
```
Comment (jeongchanKim) [inline apps/examples/ble_perfs/ble_perfs_main.c:179]: Please add space before and after `*`.

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+	double total_time = 0;
+	double speed = 0;
+
+	BLE_LOGD( "======== Bandwidth Analysis =================\n");
+        BLE_LOGD( "start time: %ld sec %ld usec \n", g_perf_data.start_time.tv_sec, g_perf_data.start_time.tv_usec);
+        BLE_LOGD( "end time: %ld sec %ld usec \n", g_perf_data.end_time.tv_sec, g_perf_data.end_time.tv_usec);
+
+	diff_time.tv_sec = g_perf_data.end_time.tv_sec - g_perf_data.start_time.tv_sec;
+	diff_time.tv_usec = g_perf_data.end_time.tv_usec - g_perf_data.start_time.tv_usec;
+	if(diff_time.tv_usec < 0) {
```
Comment (jeongchanKim) [inline apps/examples/ble_perfs/ble_perfs_main.c:166]: Please add space after `if`.

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 6]
```diff
@@ -0,0 +1,452 @@
+/****************************************************************************
+ *
+ * Copyright 2021 Samsung Electronics All Rights Reserved.
```
Comment (jeongchanKim) [inline apps/examples/ble_perfs/ble_perfs_main.c:3]: Please change to 2022.

File: apps/examples/ble_perfs/README.txt [hunk 7]
```diff
@@ -0,0 +1,8 @@
+examples/ble_rmc
+^^^^^^^^^^^^^^
+
+  This is BLE remote controller example.
```
Comment (jeongchanKim) [inline apps/examples/ble_perfs/README.txt:4]: Please check the description.
Isn't it the performance test?

File: apps/examples/ble_perfs/README.txt [hunk 8]
```diff
@@ -0,0 +1,8 @@
...
+
+  Configs (see the details on Kconfig):
+  * CONFIG_EXAMPLES_BLE_RMC
```
Comment (sangam-swami) [inline apps/examples/ble_perfs/README.txt:7]: here as well is it CONFIG_EXAMPLES_BLE_PERFS?

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 9]
```diff
@@ -0,0 +1,452 @@
...
+	return 0;
+}
+
+static int perfs_ble_parse_arguments(int argc, char **argv)
+{
+	BLE_LOGD( "parse arguments \n");
+
+	int ret = 0;
+
+	if (argc < 2) {
```
Comment (sangam-swami) [inline apps/examples/ble_perfs/ble_perfs_main.c:394]: you can put this condition above ret = 0;

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 10]
```diff
@@ -0,0 +1,452 @@
...
+		return -1;
+	}
+
+	if (strncmp(argv[1], "start", 4) == 0) {
```
Comment (sangam-swami) [inline apps/examples/ble_perfs/ble_perfs_main.c:398]: any reason for 4 as string length check

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+		ret = perfs_ble_get_result();
+		if (ret < 0) {
+			BLE_LOGD( "Fail to get performance result \n");
+			return -1;
+		}
+	} else if (strncmp(argv[1], "stop", 4) == 0 ) {
+		BLE_LOGD( "GATT server deinit \n");	
+		ret = perfs_ble_deinit();
+		if (ret < 0) {
+			BLE_LOGD( "Failed to stop esult \n");
```
Comment (sangam-swami) [inline apps/examples/ble_perfs/ble_perfs_main.c:424]: *performance result

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 12]
```diff
@@ -0,0 +1,452 @@
...
+	diff_time.tv_sec = g_perf_data.end_time.tv_sec - g_perf_data.start_time.tv_sec;
+	diff_time.tv_usec = g_perf_data.end_time.tv_usec - g_perf_data.start_time.tv_usec;
+	if(diff_time.tv_usec < 0) {
+		diff_time.tv_sec = diff_time.tv_sec - 1;
+		diff_time.tv_usec = diff_time.tv_usec + 1000000;
+	}
+
+        BLE_LOGD( "diff time: %ld sec %ld usec \n", diff_time.tv_sec, diff_time.tv_usec);
+	
+	total_time = diff_time.tv_sec + (double) diff_time.tv_usec / 1000000;
```
Comment (thapav) [inline apps/examples/ble_perfs/ble_perfs_main.c:173]: Please remove the empty space after type casting.

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 13]
```diff
@@ -0,0 +1,452 @@
...
+		p_data->packet_size = blue_data.length;
+        	gettimeofday(&p_data->start_time, NULL);
+        	BLE_LOGD( "start time: %ld sec %ld usec \n", p_data->start_time.tv_sec, p_data->start_time.tv_usec);
+	}
+
+	p_data->packet_count ++;
+
+	if (type == BLE_SERVER_ATTR_CB_WRITING_NO_RSP || type == BLE_SERVER_ATTR_CB_WRITING) {
+		if(blue_data.data[0] == blue_data.data[p_data->packet_size - 1]) {
+			//BLE_LOGD( "[WRITE] Success [%d] : %d \n" , blue_data.length, p_data->packet_count);
```
Comment (thapav) [inline apps/examples/ble_perfs/ble_perfs_main.c:210]: Is this comment not needed ?

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 14]
```diff
@@ -0,0 +1,452 @@
...
+
+		/* end time on receive of packet */
+	       	gettimeofday(&p_data->end_time, NULL);
+
+		perfs_ble_get_result();
+	}
+}
+
+static ble_server_gatt_t gatt_profile[] = {
+	{.type = BLE_SERVER_GATT_SERVICE, .uuid = {0x12,0xB6,0x6E,0x45,0xA7,0x68,0x9D,0x8D,0x9A,0x40,0x17,0x2B,0xE9,0xCB,0xF2,0x13}, .uuid_length = 16,
```
Comment (thapav) [inline apps/examples/ble_perfs/ble_perfs_main.c:231]: Add spaces after values please

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 15]
```diff
@@ -0,0 +1,452 @@
...
+		.attr_handle = BLE_APP_HANDLE_SERVICE_0,},
+
+	{.type = BLE_SERVER_GATT_CHARACT, .uuid = {0x99,0xC7,0xAA,0xE7,0xF8,0x9A,0xCB,0x88,0x43,0x4C,0x44,0xCF,0x0D,0x5B,0xDA,0xF2}, .uuid_length = 16,
```
Comment (thapav) [inline apps/examples/ble_perfs/ble_perfs_main.c:234]: same as above

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 16]
```diff
@@ -0,0 +1,452 @@
...
+	if (ble_ret != BLE_MANAGER_SUCCESS) {
+		if (ble_ret == BLE_MANAGER_ALREADY_WORKING) {
+			BLE_LOGD( "ble init is already done \n");
+		} 
+		BLE_LOGD( "Fail to init ble manager \n");
+		return -1;
+	}
+
+	g_is_ble_serv_init = 1;
+	usleep(100 * 1000);
```
Comment (thapav) [inline apps/examples/ble_perfs/ble_perfs_main.c:320]: May be we can add the sleep seconds to a #define 
#define USLEEP_USEC

File: apps/examples/ble_perfs/ble_perfs_main.c (hunk 3 above)
Comment (thapav) [reply apps/examples/ble_perfs/ble_perfs_main.c:374]: Do you mean initialize ?

File: apps/examples/ble_perfs/Make.defs (hunk 1 above)
Comment (amogh-samsung) [inline apps/examples/ble_perfs/Make.defs:3]: Could you change this file signature to say "2022" please?

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 17]
```diff
@@ -0,0 +1,452 @@
...
+		perfs_ble_reset_data();	
+		p_data->packet_size = blue_data.length;
+        	gettimeofday(&p_data->start_time, NULL);
+        	BLE_LOGD( "start time: %ld sec %ld usec \n", p_data->start_time.tv_sec, p_data->start_time.tv_usec);
+	}
+
+	p_data->packet_count ++;
+
+	if (type == BLE_SERVER_ATTR_CB_WRITING_NO_RSP || type == BLE_SERVER_ATTR_CB_WRITING) {
+		if(blue_data.data[0] == blue_data.data[p_data->packet_size - 1]) {
```
Comment (amogh-samsung) [inline apps/examples/ble_perfs/ble_perfs_main.c:209]: Please leave a charcter space after "is", "for", "while" etc

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 18]
```diff
@@ -0,0 +1,452 @@
...
+#define BLE_APP_HANDLE_CHAR_RMC_KEY (0x0074)
+#define BLE_APP_HANDLE_DESC_RMC_KEY (0x0076)
+#define BLE_APP_HANDLE_SERVICE_2 (0x0077)
+#define BLE_APP_HANDLE_CHAR_RMC_SYNC (0x0078)
+#define BLE_MAX_MTU 240
+#define MAX_PACKET_COUNT 100
+#define MAX_PACKET_SIZE BLE_MAX_MTU - 3; // 237 bytes
+#define FIRST_PACKET 11
+#define NEXT_PACKET 55
+#define LAST_PACKET 99
```
Comment (amogh-samsung) [inline apps/examples/ble_perfs/ble_perfs_main.c:49]: Could you alignt these macro definitions in the way that it's been done for some other files in this repository?

File: apps/examples/ble_perfs/Make.defs (hunk 1 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/Make.defs:3]: done

File: apps/examples/ble_perfs/Makefile (hunk 2 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/Makefile:3]: done

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 19]
```diff
@@ -0,0 +1,452 @@
...
+	}
+
+	if (strncmp(argv[1], "start", 4) == 0) {
+		BLE_LOGD( "GATT server init \n");
+		ret = perfs_ble_server_init();
+		if (ret < 0) {
+			BLE_LOGD( "Fail to init \n");
+			return -1;
+		}
+	} else if (strncmp(argv[1], "adv", 5) == 0) {
```
[the remark this answers is not in the document]
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:405]: done

File: apps/examples/ble_perfs/ble_perfs_main.c (hunk 4 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:179]: done

File: apps/examples/ble_perfs/ble_perfs_main.c (hunk 5 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:166]: done

File: apps/examples/ble_perfs/ble_perfs_main.c (hunk 6 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:3]: done

File: apps/examples/ble_perfs/README.txt (hunk 7 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/README.txt:4]: done

File: apps/examples/ble_perfs/README.txt (hunk 8 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/README.txt:7]: done

File: apps/examples/ble_perfs/ble_perfs_main.c (hunk 9 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:394]: done

File: apps/examples/ble_perfs/ble_perfs_main.c (hunk 11 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:424]: done

File: apps/examples/ble_perfs/ble_perfs_main.c (hunk 12 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:173]: done

File: apps/examples/ble_perfs/ble_perfs_main.c [hunk 20]
```diff
@@ -0,0 +1,452 @@
...
+
+
+static char * attr_cb_type_str[] = {
+	"BLE_SERVER_ATTR_CB_WRITING",
+	"BLE_SERVER_ATTR_CB_READING",
+	"BLE_SERVER_ATTR_CB_WRITING_NO_RSP",
+};
+
+static char * connection_type_str[] = {
+	"BLE_SERVER_LL_CONNECTED",
```
[the remark this answers is not in the document]
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:78]: done

File: apps/examples/ble_perfs/ble_perfs_main.c (hunk 10 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:398]: done

File: apps/examples/ble_perfs/ble_perfs_main.c (hunk 13 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:210]: done

File: apps/examples/ble_perfs/ble_perfs_main.c (hunk 14 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:231]: done

File: apps/examples/ble_perfs/ble_perfs_main.c (hunk 15 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:234]: done

File: apps/examples/ble_perfs/ble_perfs_main.c (hunk 16 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:320]: done

File: apps/examples/ble_perfs/ble_perfs_main.c (hunk 17 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:209]: done

File: apps/examples/ble_perfs/ble_perfs_main.c (hunk 18 above)
Comment (amitpurwar) [reply apps/examples/ble_perfs/ble_perfs_main.c:49]: I did not get. Please provide the reference

Comment (hdotkashyap) [reply apps/examples/ble_perfs/ble_perfs_main.c:49]: I think he meant 

```
#define abcd       1
#define pqr        2
#define lmnopq     3
```

**Review by an4967 — APPROVED** [review]
It works well.

**Pull request merged**